### PR TITLE
support restart server

### PR DESCRIPTION
**Pull request by zerothZone**
1. split shutdown to shutdown_listener and shutdown_connection
2. add start_listen

Signed-off-by: quanweiZhou <quanweiZhou@linux.alibaba.com>

### CODE REVIEW

File: src/sync/server.rs [hunk 1]
```diff
@@ -251,14 +253,14 @@ fn check_method_handler_threads(ts: &ThreadS) {
 
 impl Default for Server {
     fn default() -> Self {
-        let (rfd, wfd) = pipe2(OFlag::O_CLOEXEC).unwrap();
         Server {
             listeners: Vec::with_capacity(1),
-            monitor_fd: (rfd, wfd),
+            monitor_fd: (0, 0),
```
Comment (Tim-Zhang) [inline src/sync/server.rs:258]: 0 is not a good default value for file descriptions

File: src/sync/server.rs [hunk 2]
```diff
@@ -360,11 +355,25 @@ impl Server {
...
+                reaper_tx
+            }
+        };
 
+        let handler = thread::Builder::new()
+            .name("listener_loop".into())
+            .spawn(move || {
                 loop {
                     if service_quit.load(Ordering::SeqCst) {
+                        info!("service quit shutdown");
```
Comment (Tim-Zhang) [inline src/sync/server.rs:375]: Ambiguous and grammatical errors

File: src/sync/server.rs [hunk 3]
```diff
@@ -394,14 +404,17 @@ impl Server {
                     }
 
                     if service_quit.load(Ordering::SeqCst) {
+                        info!("service quit shutdown");
```
Comment (Tim-Zhang) [inline src/sync/server.rs:406]: Ambiguous and grammatical errors

File: src/sync/server.rs [hunk 4]
```diff
@@ -473,38 +485,68 @@ impl Server {
...
                 // notify reaper thread to exit.
                 drop(reaper_tx);
-                reaper.join().unwrap();
                 info!("ttrpc server stopped");
             })
             .unwrap();
 
         self.handler = Some(handler);
-
+        info!("server listen started");
```
Comment (Tim-Zhang) [inline src/sync/server.rs:492]: grammatical errors

File: src/sync/server.rs [hunk 5]
```diff
@@ -473,38 +485,68 @@ impl Server {
...
+        info!("close monitor");
+        if let Some(handler) = self.handler.take() {
+            handler.join().unwrap();
+        }
+        info!("join listen");
+        self
+    }
+
+    pub fn shutdown_connection(mut self) {
+        info!("shutdown connection begin");
```
Comment (Tim-Zhang) [inline src/sync/server.rs:529]: Is `start to shutdown connection` better?

File: src/sync/server.rs (hunk 1 above)
Comment (zerothZone) [reply src/sync/server.rs]: fixed

File: src/sync/server.rs (hunk 2 above)
Comment (zerothZone) [reply src/sync/server.rs]: fixed

File: src/sync/server.rs (hunk 3 above)
Comment (zerothZone) [reply src/sync/server.rs]: fixed

File: src/sync/server.rs (hunk 5 above)
Comment (zerothZone) [reply src/sync/server.rs]: fixed

**Review by lifupan — APPROVED** [review]

**Review by Tim-Zhang — APPROVED** [review]

**Pull request merged**